polkadot/node/core/pvf/prepare-worker/src/lib.rs

```diff
@@ -229,31 +234,74 @@ pub fn worker_entrypoint(
 					},
 				};
 
-				// SAFETY: new process is spawned within a single threaded process. This invariant
-				// is enforced by tests.
-				let result = match unsafe { nix::unistd::fork() } {
-					Err(errno) => Err(error_from_errno("fork", errno)),
-					Ok(ForkResult::Child) => {
-						// Dropping the stream closes the underlying socket. We want to make sure
-						// that the sandboxed child can't get any kind of information from the
-						// outside world. The only IPC it should be able to do is sending its
-						// response over the pipe.
-						drop(stream);
-						// Drop the read end so we don't have too many FDs open.
-						drop(pipe_reader);
-
-						handle_child_process(
-							pvf,
-							pipe_writer,
-							preparation_timeout,
-							prepare_job_kind,
-							executor_params,
-						)
-					},
-					Ok(ForkResult::Parent { child }) => {
-						// the read end will wait until all write ends have been closed,
-						// this drop is necessary to avoid deadlock
-						drop(pipe_writer);
+				cfg_if::cfg_if! {
+					if #[cfg(any(target_os = "linux"))] {
+						let stack = 1024 * 1024;
+						let mut stack: Vec<u8> = vec![0u8; stack];
+						let flags = CloneFlags::CLONE_NEWCGROUP
+						| CloneFlags::CLONE_NEWIPC
+						| CloneFlags::CLONE_NEWNET
+						| CloneFlags::CLONE_NEWNS
+						| CloneFlags::CLONE_NEWPID
+						| CloneFlags::CLONE_NEWUSER
+						| CloneFlags::CLONE_NEWUTS
+						| CloneFlags::from_bits_retain(libc::SIGCHLD);
```

Comment:
I'm confused by this line.

Reply:
I added a comment about this for this flag, but I removed it at some point, don't know why. I'm going to add this back.

Reply:
>  The SIGCHILD flag is needed for the waitpid, if I don't pass it the waitpid call return ECHILD

Oh I see! Yes, please add a comment, it is pretty esoteric!

Reply:
> I cannot drop either of them, because the FnMut will reuse this.

It compiles fine for me if I drop them. 🤔

Reply:
For me running `cargo test --target x86_64-unknown-linux-gnu` it doesn't work but with `cargo build --target x86_64-unknown-linux-gnu` it does.

Reply:
Both are working fine for you? Which one should I believe in? 🤣

Reply:
What platform are you on? If you're on Mac, tests are not going to work if you build for Linux. What's the error you get?

Reply:
```rust
error[E0507]: cannot move out of `stream`, a captured variable in an `FnMut` closure
   --> polkadot/node/core/pvf/prepare-worker/src/lib.rs:254:14
    |
210 |         |mut stream, worker_dir_path| {
    |          ---------- captured outer variable
...
253 |                             Box::new(|| {
    |                                      -- captured by this `FnMut` closure
254 |                                 drop(stream);
    |                                      ^^^^^^ move occurs because `stream` has type `UnixStream`, which does not implement the `Copy` trait

error[E0507]: cannot move out of `pipe_reader`, a captured variable in an `FnMut` closure
   --> polkadot/node/core/pvf/prepare-worker/src/lib.rs:255:14
    |
226 |                 let (pipe_reader, mut pipe_writer) = os_pipe::pipe()?;
    |                      ----------- captured outer variable
...
253 |                             Box::new(|| {
    |                                      -- captured by this `FnMut` closure
254 |                                 drop(stream);
255 |                                 drop(pipe_reader);
    |                                      ^^^^^^^^^^^ move occurs because `pipe_reader` has type `PipeReader`, which does not implement the `Copy` trait

error[E0382]: borrow of moved value: `stream`
   --> polkadot/node/core/pvf/prepare-worker/src/lib.rs:215:28
    |
210 |         |mut stream, worker_dir_path| {
    |          ---------- move occurs because `stream` has type `UnixStream`, which does not implement the `Copy` trait
...
215 |                 let pvf = recv_request(&mut stream)?;
    |                                        ^^^^^^^^^^^ value borrowed here after move
...
253 |                             Box::new(|| {
    |                                      -- value moved into closure here, in previous iteration of loop

error[E0382]: use of moved value: `pipe_reader`
   --> polkadot/node/core/pvf/prepare-worker/src/lib.rs:270:10
    |
226 |                 let (pipe_reader, mut pipe_writer) = os_pipe::pipe()?;
    |                      ----------- move occurs because `pipe_reader` has type `PipeReader`, which does not implement the `Copy` trait
...
253 |                             Box::new(|| {
    |                                      -- value moved into closure here
254 |                                 drop(stream);
255 |                                 drop(pipe_reader);
    |                                      ----------- variable moved due to use in closure
...
270 |                                     pipe_reader,
    |                                     ^^^^^^^^^^^ value used here after move```

Reply:
This makes sense to me, cause if stream and pipe are being moved to the callback they won't be available for the parent, right?